[NCL-7594] add back pressure on 425

michalovjan opened the pull request.

thescouser89 commented on core/src/test/java/org/jboss/pnc/rex/core/endpoints/HttpEndpoint.java:
```diff
@@ -106,6 +111,27 @@ public Response acceptAndFail(StartRequest request) {
+        counter.incrementAndGet();
+        return Response.status(425)
+                .header("Content-Type", "application/json")
+                .entity("{\"you\":\"wait\"}")
```
:no_good:

thescouser89 reviewed (approved):
While I appreciate your faith on me, I don't think I have the skills to review this PR. :(

I do have 1 question though: is this for the rex-client codebase to talk to Rex, or is this for when Rex sends a request to a service and the service replies back with 425?

Is the HTTP 409 tests also related to the HTTP 425 or are they unrelated?

michalovjan commented:
> While I appreciate your faith on me, I don't think I have the skills to review this PR. :(
DW, Atleast someone have seen it! :)

I'll write a summary adding the context.
 
> I do have 1 question though: is this for the rex-client codebase to talk to Rex, or is this for when Rex sends a request to a service and the service replies back with 425?

Context:
This is solely for HTTP client that Rex uses to communicate with _other services_ (+ test endpoints), so request to start/stop/notify. With Matej we've found a HTTP code 425 Too Early [1] that we wanted to use for backpressure (although Mozilla states a different use-case, we liked the name of it) Response of 4XX usually means a failure, therefore Rex would transition a task to START_FAILED or STOP_FAILED but 425 is special... Basically, if ANY service that Rex communicates with responds with 425, Rex will retry after some time (exponentially increasing) and the task can still end up as UP or STOPPING.

Responding with 425 means that the request came too early and Rex should wait a bit. For example in Orch we've found an edge case where a Group Build wasn't yet commited into DB and Rex sent a notification to Orch which required the GB to be present. Also, this can solve stuff in general when we have services that are currently fully occupied (OCP pod resources). Maybe this could somehow also be beneficial with RHPAM replacement. 

[1] https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/425

> Is the HTTP 409 tests also related to the HTTP 425 or are they unrelated?

Uff, unrelated. I think I've named the test endpoint 409xxxxx. I'll fix that. 409s happen when someone tries to schedule a task with the same name twice, whilst 425 is for backpressure. 

Additional stuff done:

1. Backpressure is configurable. Basically, you can configure delays, exponential backoff maxs and mins, and choose whether to use a fixed amount of retries for a request like 20 times or 
2. Changed naming of some configuration because it was starting to get vague. Internal FT for transactions vs FT for HTTP requests.
3. Refactored the HTTP client a bit.
4. Fixed a bug where Rex wouldn't apply query parameters for HTTP requests.

thescouser89 commented:
Sounds good to me!

Ship it! :rocket: :rocket: :rocket:
